content/docs/reference/design-docs/signaling.en.md

```diff
@@ -0,0 +1,419 @@
+   stopsAtBufferStop: bool,
+   signalTypes: Vec<SignalingSystemId>,
+   signalSettings: Vec<SignalSettings>,
+   signalPositions: Vec<Distance>,
+   length: Distance,
+}
+
+/// Runs in the signaling system module
+fn check_block(
+   block: Block,
+);
+
+
+/// Runs in the signal driver module
+fn check_signal(
+   signal: SignalSettings,
+   block: Block, // The partial block downstream of the signal - no signal can see backward
+);
+```
+
+### Signal lifecycle
+
+Before a train startup:
+
+- the path a of the train can be expressed as routes or as blocks. Those two paths should overlap.
+- the signal queue a train will encounter is established.
+
+During the simulation:
+- along a train movement, the track occupation before it are synthesized
+- when a train observes a signal, its state is evaluated
+
+### Signal state evaluation
+
+Signals are modeled as an evaluation function, taking a view of the world and returning the signal state
+
+```kotlin
+
+enum ZoneStatus {
+   /** The zone is clear to be used by the train */
+   CLEAR,
+   /** The zone is occupied by another train, but otherwise clear to use */
+   OCCUPIED,
+   /** The zone is incompatible. There may be another train as well */
+   INCOMPATIBLE,
+}
+
+interface MAView {
+    /** Combined status of the zones protected by the current signal */
+    val protectedZoneStatus: ZoneStatus
+    val nextSignalState: SignalState
+    val nextSignalSettings: SignalSettings
+}
+
+interface DirectSpeedLimit {
+    /** Distance between the signal and the speed limit */
+    val distance: Distance
+    val speed: Speed
+}
+
+interface IndirectSpeedLimit {
+    val distanceToNextSignal: Distance
+    val nextSignalState: SignalState
+    val nextSignalSettings: SignalSettings
+}
+
+interface SpeedLimitView {
+    /** A list of speed limits directly downstream of the signal */
+    val directSpeedLimits: List<DirectSpeedLimit>
+    /** A list of speed limits which need to be announced in a signal chain */
+    val indirectSpeedLimits: List<IndirectSpeedLimit>
+}
+
+fun signal(maView: MAView?, limitView: SpeedLimitView?): SignalState {
+    // ...
+}
+```
+
+The view should allow access to the following data:
+ - a synthetized view of zones downstream until the end of the train's MA
+ - the block chain
+ - the state of downstream signals which belong to the current block chain
+
+### Signaling view path
+
+The path along which the MAView and SpeedLimitView live is best expressed using blocks:
+
+- blocks can be added to extend the view along the path of a train
+- the view can be reduced by removing blocks, as the train passes by signals
+
+### Simulation outside the train path
+
+It is possible to simulate signaling outside of any train path:
+
+   - if a signal starts blocks which have differing paths, it is simulated as if it were at the end of a route
+   - if a signal starts blocks which all start the same path, it is simulated in the same view as the next signals in this path
```

Comment:
I didn't really undertand this part

Reply:
It's about simulating signals a rest. At some point there was a proposed feature where we might have need to do that.